tracer/src/Datadog.Trace/Activity/ActivityListener.cs

```diff
@@ -0,0 +1,294 @@
+// <copyright file="ActivityListener.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Threading;
+using System.Threading.Tasks;
+using Datadog.Trace.Activity.DuckTypes;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Activity
+{
+    internal static class ActivityListener
+    {
+        private const int InitializationBackoffPerRetry = 10000;
+
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(ActivityListener));
+
+        private static int _initializationRetries = 5;
+
+        private static object? _activityListenerInstance;
+        private static Func<object>? _getCurrentActivity;
+
+        private static int _initialized = 0;
+        private static int _stopped = 0;
+
+        public static bool IsRunning
+        {
+            get
+            {
+                return Interlocked.CompareExchange(ref _initialized, 1, 1) == 1 &&
+                       Interlocked.CompareExchange(ref _stopped, 0, 0) == 0;
+            }
+        }
+
+        internal static object? GetCurrentActivityObject()
+        {
+            try
+            {
+                return _getCurrentActivity?.Invoke();
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error calling Activity.Current");
+            }
+
+            return null;
+        }
+
+        public static IActivity? GetCurrentActivity()
+        {
+            var activity = GetCurrentActivityObject();
+
+            if (activity is null)
+            {
+                return null;
+            }
+
+            if (activity.TryDuckCast<IActivity6>(out var activity6))
+            {
+                return activity6;
+            }
+
+            if (activity.TryDuckCast<IActivity5>(out var activity5))
+            {
+                return activity5;
+            }
+
+            if (activity.TryDuckCast<IW3CActivity>(out var w3cActivity))
+            {
+                return w3cActivity;
+            }
+
+            return activity.TryDuckCast<IActivity>(out var activity4) ? activity4 : null;
```

Comment:
Could you chain `DuckAs<T>()` with null coalescing? Something like:
```csharp
return activity.DuckAs<IActivity6>() ??
       activity.DuckAs<IActivity5>() ??
       activity.DuckAs<IW3CActivity>() ??
       activity.DuckAs<IActivity>();
```

Level of care low, just curious.

Reply:
I hadn't thought about it :D done in: https://github.com/DataDog/dd-trace-dotnet/pull/2446/commits/35b9dbdffa19d1d550522023f8e01d53b8f8adb3